File: server/src/main/java/dev/slimevr/poserecorder/BVHRecorder.java
package dev.slimevr.poserecorder;

import dev.slimevr.VRServer;
import dev.slimevr.posestreamer.BVHFileStream;
import dev.slimevr.posestreamer.PoseDataStream;
import dev.slimevr.posestreamer.ServerPoseStreamer;
import io.eiren.util.logging.LogManager;

import java.io.File;
import java.io.IOException;


public class BVHRecorder {

	private static final File bvhSaveDir = new File("BVH Recordings");
	private final ServerPoseStreamer poseStreamer;
	private PoseDataStream poseDataStream = null;

	public BVHRecorder(VRServer server) {
		this.poseStreamer = new ServerPoseStreamer(server);
	}

	public void startRecording() {
		File bvhFile = getBvhFile();
		if (bvhFile != null) {
			try {
				poseDataStream = new BVHFileStream(bvhFile);
				poseStreamer.setOutput(poseDataStream, 1000L / 100L);
			} catch (IOException e1) {
				LogManager
					.severe(
						"[BVH] Failed to create the recording file \"" + bvhFile.getPath() + "\"."
					);
			}
		} else {
			LogManager.severe("[BVH] Unable to get file to save to");
		}
	}

	public void endRecording() {

		try {
			poseStreamer.closeOutput(poseDataStream);
		} catch (Exception e1) {
			LogManager.severe("[BVH] Exception while closing poseDataStream", e1);
		} finally {
			poseDataStream = null;
		}
	}

	private File getBvhFile() {
		if (bvhSaveDir.isDirectory() || bvhSaveDir.mkdirs()) {
			File saveRecording;
			int recordingIndex = 1;
			do {
				saveRecording = new File(bvhSaveDir, "BVH-Recording" + recordingIndex++ + ".bvh");
			} while (saveRecording.exists());

			return saveRecording;
		} else {
			LogManager
				.severe(
					"[BVH] Failed to create the recording directory \""
						+ bvhSaveDir.getPath()
						+ "\"."
				);
		}

		return null;
	}

	public boolean isRecording() {
		return poseDataStream != null;
	}
}

File: server/core/src/main/java/dev/slimevr/posestreamer/BVHRecorder.kt
package dev.slimevr.posestreamer

import dev.slimevr.VRServer
import io.eiren.util.logging.LogManager
import java.io.File
import java.io.IOException
import java.nio.file.Path

class BVHRecorder(server: VRServer) {
	private val poseStreamer: ServerPoseStreamer = ServerPoseStreamer(server)
	private var poseDataStream: PoseDataStream? = null

	val isRecording: Boolean
		get() = poseDataStream != null

	fun startRecording(path: Path) {
		val filePath = path.toFile()

		val file = if (filePath.isDirectory()) {
			getBvhFile(filePath) ?: return
		} else {
			filePath
		}

		try {
			val stream = BVHFileStream(file)
			poseDataStream = stream
			// 100 FPS
			poseStreamer.setOutput(stream, 1f / 100f)
		} catch (_: IOException) {
			LogManager.severe("[BVH] Failed to create the recording file \"${file.path}\".")
		}
	}

	fun endRecording() {
		try {
			val stream = poseDataStream
			if (stream != null) {
				poseStreamer.closeOutput(stream)
			}
		} catch (e1: Exception) {
			LogManager.severe("[BVH] Exception while closing poseDataStream", e1)
		} finally {
			poseDataStream = null
		}
	}

	private fun getBvhFile(bvhSaveDir: File): File? {
		if (bvhSaveDir.isDirectory() || bvhSaveDir.mkdirs()) {
			var saveRecording: File?
			var recordingIndex = 1
			do {
				saveRecording =
					File(bvhSaveDir, "BVH-Recording${recordingIndex++}.bvh")
			} while (saveRecording.exists())

			return saveRecording
		} else {
			LogManager
				.severe(
					"[BVH] Failed to create the recording directory \"${bvhSaveDir.path}\".",
				)
		}

		return null
	}
}
